### PR TITLE
Muon - Algorithm; MuonGroupingCounts

### PR DESCRIPTION
**Pull request by TWJubb**
Refs #23642

Requires #23677, #23678 to be merged.

### Summary

This PR introduces a new algorithm `MuonGroupingCounts` which is one of a set of four algorithms intended to break up the large `MuonProcess` algorithm which is not fit for purpose with the user requirements of the new muon interface (it is too large and does not allow easy or intuitive scripting of the user workflows).

This algorithm captures the idea of a "grouping counts" operation on a workspace; which is essentially a summation of selected spectra. The input to this algorithm is intended to come from the output of `MuonPreProcess` (#23678).

The reason for this is to treat multi-period data in a consistent way; by default a single period workspace will be converted to a workspace group with a single workspace by `MuonPreProcess`; and this is what `MuonGroupingCounts` expects.

Creating this algorithm creates a logically distinct operation which is well understood by users familiar with muon workflows; rather than having the grouping counts as a subset of the behaviour of one gigantic algorithm whose >10 inputs are not intuitive. The reason to have it this way is to fulfil the user requirement of being able to replicate the muon interface workflow in the python scripting window, if required.

muongroupingcounts

### Unit Tests

As per instruction; unit tests have not been written yet, to focus on the functionality.

-----------------------------------------

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by Mantid-Matthew** [comment]
The failing doctests here are due to the fact that #23678 needs to be merged first

File: Framework/Muon/inc/MantidMuon/MuonGroupingCounts.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+  const std::string name() const override { return "MuonGroupingCounts"; }
+  int version() const override { return (1); }
+  const std::string category() const override { return "Muon\\DataHandling"; }
+  const std::string summary() const override {
+    return "Apply a grouping (summation of counts) across a set of detectors "
+           "in Muon data.";
+  }
+  const std::vector<std::string> seeAlso() const override {
+    return {"MuonProcess"};
```
Comment (AnthonyLim23) [inline Framework/Muon/inc/MantidMuon/MuonGroupingCounts.h]: should include child algs and muonPreProcess and other new muon algs (later)

File: Framework/Muon/src/MuonGroupingCounts.cpp [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+using namespace Mantid::HistogramData;
+
+namespace {
+
+bool checkPeriodInWorkspaceGroup(const int &period,
+                                 WorkspaceGroup_sptr workspace) {
+  return period <= workspace->getNumberOfEntries();
+}
+
+bool is_alnum_underscore(char c) {
```
Comment (AnthonyLim23) [inline Framework/Muon/src/MuonGroupingCounts.cpp]: what is alnum?

Comment (AnthonyLim23) [reply Framework/Muon/src/MuonGroupingCounts.cpp]: also could we have a longer variable name (single character variables are prone to typos)

File: Framework/Muon/src/MuonGroupingCounts.cpp [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+MatrixWorkspace_sptr groupDetectors(MatrixWorkspace_sptr workspace,
+                                    const std::vector<int> &detectorIDs) {
+
+  auto outputWS = WorkspaceFactory::Instance().create(workspace, 1);
+
+  std::vector<size_t> wsIndices =
+      workspace->getIndicesFromDetectorIDs(detectorIDs);
+
+  if (wsIndices.size() != detectorIDs.size())
+    throw std::invalid_argument("Some of the detector IDs were not found");
```
Comment (AnthonyLim23) [inline Framework/Muon/src/MuonGroupingCounts.cpp]: can we change this error message to something like: "Requested a detector outside of range ( wsIndicies.size() > detectorIDs.size)"?

File: Framework/Muon/test/MuonGroupingCountsTest.h [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+
+  void
+  test_that_cannot_add_spectra_to_group_which_exceed_those_in_the_workspace() {
+    auto ws = createMultiPeriodWorkspaceGroup(1, 5, 10, "group1");
+
+    std::vector<int> detectors = {6, 7, 8, 9, 10};
+    auto alg = setUpAlgorithmWithGroupNameAndDetectors(ws, "group1", detectors);
+
+    alg->execute();
+    TS_ASSERT(!alg->isExecuted());
```
[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply Framework/Muon/test/MuonGroupingCountsTest.h]: It does throw an error on line 44 which is modified above. Algorithms by default don't rethrow errors however they instead print out the message to the dialog window.

[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply Framework/Muon/test/MuonGroupingCountsTest.h]: we would like them to throw in the GUI so we can then catch the error and put it into a warning

[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply Framework/Muon/test/MuonGroupingCountsTest.h]: I will check what ASSERT_THROWS does though as slightly uncertain here.

[the remark this answers is not in the document]
Comment (Mantid-Matthew) [reply Framework/Muon/test/MuonGroupingCountsTest.h]: We could easily add it here? https://github.com/mantidproject/mantid/blob/master/Framework/Muon/src/MuonAlgorithmHelper.cpp#L582

[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply Framework/Muon/test/MuonGroupingCountsTest.h]: Lets give it a go

[the remark this answers is not in the document]
Comment (AnthonyLim23) [reply Framework/Muon/test/MuonGroupingCountsTest.h]: dont need to do it now, just later

File: Framework/Muon/src/MuonGroupingCounts.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+                                    const std::vector<int> &detectorIDs) {
+
+  auto outputWS = WorkspaceFactory::Instance().create(workspace, 1);
+
+  std::vector<size_t> wsIndices =
+      workspace->getIndicesFromDetectorIDs(detectorIDs);
+
+  if (wsIndices.size() != detectorIDs.size())
+    throw std::invalid_argument("Requested a detector outside of range "
+                                "(wsIndicies.size() > detectorIDs.size)");
```
Comment (AnthonyLim23) [inline Framework/Muon/src/MuonGroupingCounts.cpp]: sorry, but could we have the numbers displayed instead of detectorIDs.size?

**Comment by AnthonyLim23** [comment]
needs a clang otherwise looks good

**Review by AnthonyLim23 — DISMISSED** [review]

File: Framework/Muon/CMakeLists.txt [hunk 6]
```diff
@@ -13,6 +13,7 @@ set ( SRC_FILES
         src/MuonAlgorithmHelper.cpp
```
Comment (gvardany) [inline Framework/Muon/CMakeLists.txt]: It seems the indentation gone wrong in the cmake file.

File: Framework/Muon/inc/MantidMuon/MuonAlgorithmHelper.h [hunk 7]
```diff
@@ -139,6 +139,8 @@ DLLExport void addSampleLog(Mantid::API::MatrixWorkspace_sptr workspace,
                             const std::string &logName,
                             const std::string &logValue);
 
+DLLExport bool is_alphanumeric_or_underscore(char character);
```
Comment (gvardany) [inline Framework/Muon/inc/MantidMuon/MuonAlgorithmHelper.h]: The method names must be in camelCase instead of using `_`.

**Review by gvardany — CHANGES_REQUESTED** [review]
Please see few minor comments inline.

**Review by gvardany — APPROVED** [review]

**Pull request merged**